File: frontend/lib/api/analysis.ts
import { API_URL } from "./config";
import { apiFetch } from "./errors";

export type ScoringBreakdown = {
  skill_score: number;
  keyword_score: number;
  semantic_score: number;
  role_alignment_score: number;
  evidence_score: number;
  length_sanity: number;
  confidence: "high" | "medium" | "low" | string;
  final_score: number;
  resume_role_family?: string;
  jd_role_family?: string;
  resume_role_signals?: Record<string, string[]>;
  jd_role_signals?: Record<string, string[]>;
  resume_stack_groups?: string[];
  jd_stack_groups?: string[];
  critical_skills?: string[];
  role_alignment_notes?: string[];
  evidence_notes?: string[];
};

export type PrioritizedMissingSkills = {
  high_priority: string[];
  medium_priority: string[];
  low_priority: string[];
};

export type ResumeFeedbackItem = {
  title: string;
  message: string;
  why_this_matters: string;
  suggested_edit?: string | null;
};

export type ResumeFeedback = {
  critical_gaps: ResumeFeedbackItem[];
  cv_wording_improvements: ResumeFeedbackItem[];
  suggested_bullet_rewrites: ResumeFeedbackItem[];
  missing_evidence_areas: ResumeFeedbackItem[];
  recommended_next_edits: ResumeFeedbackItem[];
};


export type SemanticInsights = {
  enabled: boolean;
  model_name?: string | null;
  resume_jd_similarity?: number | null;
  confidence: "high" | "medium" | "low" | string;
  notes: string[];
  reason?: string | null;
};

export type HybridEvaluation = {
  enabled: boolean;
  hybrid_score_candidate: number;
  rule_based_score: number;
  semantic_component?: number | null;
  taxonomy_component: number;
  confidence_adjustment: number;
  explanation_notes: string[];
  production_safe: boolean;
};

export type MLEvaluation = {
  enabled: boolean;
  predicted_label?: string | null;
  confidence?: number | null;
  label_probabilities: Record<string, number>;
  model_version?: string | null;
  production_safe: boolean;
  note: string;
  reason?: string | null;
};

export type MatchAnalysis = {
  id: number;
  user_id: number;
  resume_id: number;
  job_description_id: number;
  match_score: number;
  matched_skills: string[];
  missing_skills: string[];
  keyword_overlap: string[];
  summary: string;
  suggestions: string[];
  resume_feedback: ResumeFeedback;
  resume_text_preview: string;
  jd_text_preview: string;
  resume_detected_skills: string[];
  jd_detected_skills: string[];
  scoring_breakdown: ScoringBreakdown;
  skill_gap_summary: string;
  prioritized_missing_skills: PrioritizedMissingSkills;
  improvement_plan: string[];
  semantic_insights?: SemanticInsights;
  hybrid_evaluation?: HybridEvaluation;
  ml_evaluation?: MLEvaluation;
  created_at: string;
  updated_at: string;
};

export type ResumeJobMatchPayload = {
  resume_id: number;
  job_description_id: number;
};

async function analysisRequest<T>(path: string, token: string, init: RequestInit = {}): Promise<T> {
  const response = await apiFetch(`${API_URL}${path}`, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
      ...init.headers
    }
  }, "Kh\u00f4ng th\u1ec3 x\u1eed l\u00fd ph\u00e2n t\u00edch CV v\u00e0 JD.");

  return response.json() as Promise<T>;
}

export function runResumeJobMatch(token: string, payload: ResumeJobMatchPayload): Promise<MatchAnalysis> {
  return analysisRequest<MatchAnalysis>("/api/analysis/resume-job-match", token, {
    method: "POST",
    body: JSON.stringify(payload)
  });
}

export function getAnalysisHistory(token: string): Promise<MatchAnalysis[]> {
  return analysisRequest<MatchAnalysis[]>("/api/analysis/history", token);
}

File: backend/app/schemas/analysis.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ResumeJobMatchRequest(BaseModel):
    resume_id: int = Field(gt=0)
    job_description_id: int = Field(gt=0)


class ScoringBreakdown(BaseModel):
    skill_score: float
    keyword_score: float
    semantic_score: float
    role_alignment_score: float = 0
    evidence_score: float = 0
    length_sanity: float
    confidence: str = "medium"
    final_score: float
    resume_role_family: str = "general software"
    jd_role_family: str = "general software"
    resume_role_signals: dict[str, list[str]] = Field(default_factory=dict)
    jd_role_signals: dict[str, list[str]] = Field(default_factory=dict)
    resume_stack_groups: list[str] = Field(default_factory=list)
    jd_stack_groups: list[str] = Field(default_factory=list)
    critical_skills: list[str] = Field(default_factory=list)
    role_alignment_notes: list[str] = Field(default_factory=list)
    evidence_notes: list[str] = Field(default_factory=list)


class PrioritizedMissingSkills(BaseModel):
    high_priority: list[str]
    medium_priority: list[str]
    low_priority: list[str]


class ResumeFeedbackItem(BaseModel):
    title: str
    message: str
    why_this_matters: str
    suggested_edit: str | None = None


class ResumeFeedback(BaseModel):
    critical_gaps: list[ResumeFeedbackItem] = Field(default_factory=list)
    cv_wording_improvements: list[ResumeFeedbackItem] = Field(default_factory=list)
    suggested_bullet_rewrites: list[ResumeFeedbackItem] = Field(default_factory=list)
    missing_evidence_areas: list[ResumeFeedbackItem] = Field(default_factory=list)
    recommended_next_edits: list[ResumeFeedbackItem] = Field(default_factory=list)


class TaxonomyInsight(BaseModel):
    role_family: str = "general software"
    stack_groups: list[str] = Field(default_factory=list)
    normalized_skills: list[str] = Field(default_factory=list)
    related_skill_suggestions: list[str] = Field(default_factory=list)


class MatchTaxonomyInsights(BaseModel):
    role_family: str = "general software"
    stack_groups: list[str] = Field(default_factory=list)
    normalized_skills: list[str] = Field(default_factory=list)
    related_skill_suggestions: list[str] = Field(default_factory=list)
    resume: TaxonomyInsight = Field(default_factory=TaxonomyInsight)
    job_description: TaxonomyInsight = Field(default_factory=TaxonomyInsight)



class SemanticInsights(BaseModel):
    enabled: bool = False
    model_name: str | None = None
    resume_jd_similarity: float | None = None
    confidence: str = "low"
    notes: list[str] = Field(default_factory=list)
    reason: str | None = None


class HybridEvaluation(BaseModel):
    enabled: bool = False
    hybrid_score_candidate: float
    rule_based_score: float
    semantic_component: float | None = None
    taxonomy_component: float
    confidence_adjustment: float
    explanation_notes: list[str] = Field(default_factory=list)
    production_safe: bool = True


class MLEvaluation(BaseModel):
    enabled: bool = False
    predicted_label: str | None = None
    confidence: float | None = None
    label_probabilities: dict[str, float] = Field(default_factory=dict)
    model_version: str | None = None
    production_safe: bool = False
    note: str = "ML prediction chỉ dùng để đánh giá nội bộ, chưa thay thế điểm chính."
    reason: str | None = None

class MatchAnalysisResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    resume_id: int
    job_description_id: int
    match_score: float
    matched_skills: list[str]
    missing_skills: list[str]
    keyword_overlap: list[str]
    summary: str
    suggestions: list[str]
    resume_feedback: ResumeFeedback
    resume_text_preview: str
    jd_text_preview: str
    resume_detected_skills: list[str]
    jd_detected_skills: list[str]
    scoring_breakdown: ScoringBreakdown
    skill_gap_summary: str
    prioritized_missing_skills: PrioritizedMissingSkills
    improvement_plan: list[str]
    taxonomy_insights: MatchTaxonomyInsights = Field(default_factory=MatchTaxonomyInsights)
    semantic_insights: SemanticInsights = Field(default_factory=SemanticInsights)
    hybrid_evaluation: HybridEvaluation
    ml_evaluation: MLEvaluation = Field(default_factory=MLEvaluation)
    created_at: datetime
    updated_at: datetime

